Performance measurement

sgraband opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

This commit adds a script for performance measurement using the Chrome DevTools tracing.
Currently the startup time of the browser-app is being measured.
The script uses Puppeteer so other processes can be measured in the future as well.
For more information see the `README.md` in the scripts/performance directory.

This PR is intended as the first step to an eventual performance test suite which could also be integrated into the CI tooling.
See https://github.com/eclipse-theia/(remove this parenthesis when opening against Theia)theia/issues/9722 for more information about this topic.

Example output:
```
[Performance][Startup][1] Largest Contentful Paint (LCP): 2.95 seconds
[Performance][Startup][2] Largest Contentful Paint (LCP): 2.823 seconds
[Performance][Startup][3] Largest Contentful Paint (LCP): 2.829 seconds
[Performance][Startup][4] Largest Contentful Paint (LCP): 2.835 seconds
[Performance][Startup][5] Largest Contentful Paint (LCP): 2.778 seconds
[Performance][Startup][6] Largest Contentful Paint (LCP): 2.795 seconds
[Performance][Startup][7] Largest Contentful Paint (LCP): 2.772 seconds
[Performance][Startup][8] Largest Contentful Paint (LCP): 2.672 seconds
[Performance][Startup][9] Largest Contentful Paint (LCP): 2.706 seconds
[Performance][Startup][10] Largest Contentful Paint (LCP): 2.804 seconds
[Performance][Startup][MEAN] Largest Contentful Paint (LCP): 2.796 seconds
[Performance][Startup][STDEV] Largest Contentful Paint (LCP): 0.072 seconds
```

Contributed on behalf of STMicroelectronics

Signed-off-by: sgraband@eclipsesource.com
#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
See the README on how you can run the script manually or run the `performance:startup` node script in the root directory to measure the startup time. You can also adjust the parameters to for example execute more iterations or to run the test headful.
Additionally, you can import the generated files in `scripts/performance/profiles` in the performance tab of the Chrome DevTools.
#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

sdirix commented on scripts/performance/README.md:
```diff
@@ -0,0 +1,27 @@
+# Performance measurements
+
+This directory contains a script that measures the performance of theia.
```
```suggestion
This directory contains a script that measures the performance of Theia.
```

sdirix commented on scripts/performance/README.md:
```diff
@@ -0,0 +1,27 @@
+# Performance measurements
+
+This directory contains a script that measures the performance of theia.
+Currently the support is limited to measuring the `browser-app`s startup time using the `Largest contentful Paint (LCP)` value.
```
```suggestion
Currently the support is limited to measuring the `browser-app`s startup time using the `Largest contentful paint (LCP)` value.
```

sdirix commented on scripts/performance/README.md:
```diff
@@ -0,0 +1,27 @@
+
+### Prerequisites
+
+To run the script the theia backend needs to be started.
```
```suggestion
To run the script the Theia backend needs to be started.
```

sdirix commented on scripts/performance/README.md:
```diff
@@ -0,0 +1,27 @@
+To run the script the theia backend needs to be started.
+This can either be done with the `Launch Browser Backend` launch config or by running `yarn start` in the `examples/browser-app` directory.
+
+Additionally, the root package.json file contains node scripts that startup the backend as well as the script. Like the `performance:startup` script.
```
```suggestion
Additionally, the root package.json file contains node scripts that startup the backend as well as the script.
See the `performance:startup` script.
```

sdirix commented on scripts/performance/README.md:
```diff
@@ -0,0 +1,27 @@
+The script accepts the following parameters:
+
+-   `--name`: Specify a name for the current measurement (default: `Measurement`)
+-   `--url`: Point theia to a url for example for specifying a specifc workspace (default: `http://localhost:3000/#/<absolutePathToWorkspaceFolder>`)
```
```suggestion
-   `--url`: Point Theia to a url for example for specifying a specifc workspace (default: `http://localhost:3000/#/<absolutePathToWorkspaceFolder>`)
```

sdirix commented on scripts/performance/measure-performance.js:
```diff
@@ -0,0 +1,155 @@
+    }
+}
+
+async function measureStartup(profilePath, name, runNr, multipleRuns) {
```
* Does this need to be `async`?
 * I would rename this to something like: analyzeStartup which returns the parsed values. The function calling this one then calls `logDuration` with the result. The idea is to separate between file analysis and IO.

sdirix commented on scripts/performance/profiles/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# This folder contains all the generated performance profiles.
```
Do we need this directory? We could also just check whether it exists and create it if necessary in our script.

sdirix commented on scripts/performance/workspace/.gitignore:
```diff
@@ -0,0 +1 @@
+# This just checks in an empty workspace for the performance measurements. 
```
Do we need this directory? We could also just check whether it exists and create it if necessary in our script.

sdirix commented on scripts/performance/measure-performance.js:
```diff
@@ -0,0 +1,155 @@
+const performanceTag = braceText('Performance');
+
+let name = 'Measurement';
+let url = 'http://localhost:3000/#' + path;
```
I would just handin the whole url via a parameter. then this script can be used with arbitrary Theia instances.

sgraband replied:
If you pass a URL as a parameter you need to specify the whole URL. This is just done, to have a relative path to the workspace for the default value.

sdirix replied:
Wouldn't this be more flexible? Now it's only possible to test locally hosted Theia. With the full URL you can even test deployed Theias. Maybe use url if it's given (default to `http://localhost:3000/#`) and append the `path` when given

sgraband replied:
If you pass anything with the `--url` parameter the url will be completly overwritten. Hence, you could just call the script with your hosted url. However with this you cannot just append a workspace to the `http://localhost:3000/#`. Should i split the arguments into domain and workspace?

sdirix replied:
Ah I see, I did not recognize that `url` was already a parameter. But if the `url` is given we should probably not call `getOrCreateFolder(path)`, right?

sgraband replied:
Good Catch! I will add a check whether or not a url has been passed.

sdirix commented on package.json:
```diff
@@ -76,7 +77,9 @@
     "debug:electron": "yarn rebuild:electron && yarn --cwd examples/electron start:debug",
-    "download:plugins": "theia download:plugins"
+    "download:plugins": "theia download:plugins",
+    "performance:startup": "start-server-and-test start:browser 3000 performance:startup:script",
```
Do we have to wait for the Theia app to start before executing the performance script? We could also do that check in our script ourselves, right? Then we don't need the dependency and only need to add one script here.

sdirix commented on scripts/performance/measure-performance.js:
```diff
@@ -0,0 +1,155 @@
+    if (multipleRuns) {
+        const mean = calculateMean(durations);
+        logDuration(name, 'MEAN', lcp, mean);
+        console.log(performanceTag + braceText(name) + braceText('STDEV') + ' ' + lcp + ': ' + calculateStandardDeviation(mean, durations));
```
Missing `seconds` at the end

sdirix commented on scripts/performance/measure-performance.js:
```diff
@@ -0,0 +1,155 @@
+
+        await page.goto(url);
+        await page.waitForSelector('.theia-preload', { state: 'detached' });
+        await page.waitForSelector('.fa-exclamation-triangle', { state: 'visible' });
```
Please add a comment why this selector was chosen and what it's actually choosing

sdirix reviewed (changes requested):
Works for me!

I get this weird warning during execution:
> The node binary used for scripts is /tmp/yarn--1627025300882-0.06251617126804576/node but npm is using /home/stefan/.nvm/versions/node/v12.22.1/bin/node itself. Use the `--scripts-prepend-node-path` option to include the path for the node binary npm was
 executed with.

sdirix commented:
Please check whether you can add
```
// @ts-check
```
to the script

sgraband commented:
I updated my changes with your feedback. Regarding the error you mentioned: This should not happen anymore. (I added a npm preference, that prepends the node path)

Additionally i enabled the ` @ts-check`. There was however a small problem with the puppeteer selectors, so i ignored those lines.

sdirix commented:
The `yarn.lock` changes can be removed

sdirix commented on package.json:
```diff
@@ -76,7 +76,9 @@
     "debug:electron": "yarn rebuild:electron && yarn --cwd examples/electron start:debug",
-    "download:plugins": "theia download:plugins"
+    "download:plugins": "theia download:plugins",
+    "performance:startup": "concurrently --success first -k -r \"yarn performance:script --name Startup --folder startup --runs 2\" \"yarn start:browser\"",
```
Maybe inline the `performance:script` and remove the separate `performance script`

sgraband commented:
Removed the `yarn.lock` changes and updated the `package.json`.

sdirix commented on scripts/performance/README.md:
```diff
@@ -0,0 +1,29 @@
+
+## Running the script
+
+### Prerequisites
+
+To run the script the Theia backend needs to be started.
+This can either be done with the `Launch Browser Backend` launch config or by running `yarn start` in the `examples/browser-app` directory.
+
+Additionally, the root package.json file contains node scripts that startup the backend as well as the script.
+See the `performance:startup` script.
```
```suggestion
### Quick Start

Execute `yarn run performance:startup` in the root directory.

### Prerequisites

To run the performance tests the Theia backend needs to be started.
This can either be done with the `Launch Browser Backend` launch config or by running `yarn start` in the `examples/browser-app` directory.
```

sdirix commented on scripts/performance/README.md:
```diff
@@ -0,0 +1,29 @@
+
+The script can be exectued using `node measure-performance.js` in this directory.
+
+The script accepts the following parameters:
```
```suggestion
The script accepts the following optional parameters:
```

sdirix reviewed (changes requested):
The `profiles` folder and default workspace folder should still be ignored, however by a `.gitignore` in the `scrips/performance` directory.

sdirix commented on scripts/performance/README.md:
```diff
@@ -0,0 +1,29 @@
+The script accepts the following parameters:
+
+-   `--name`: Specify a name for the current measurement (default: `Measurement`)
+-   `--url`: Point Theia to a url for example for specifying a specifc workspace (default: `http://localhost:3000/#/<absolutePathToWorkspaceFolder>`)
```
Default is workspace next to the script, right?
```suggestion
-   `--url`: Point Theia to a url for example for specifying a specifc workspace (default: `http://localhost:3000/#/<pathToMeasurementScript>/workspace`)
```

sdirix commented on scripts/performance/measure-performance.js:
```diff
@@ -0,0 +1,186 @@
+        await page.tracing.start({ path: file, screenshots: true });
+
+        await page.goto(url);
+        // @ts-ignore
+        await page.waitForSelector('.theia-preload', { state: 'detached' });
```
Is `{ state: 'detached' }` even a valid option for Puppeteer? Seems to be Playwright only. What's the difference to just removing this option?

sgraband replied:
I just removed this selector, as it does not affect the script at all. Waiting for the other selector is enough.

sdirix commented on scripts/performance/measure-performance.js:
```diff
@@ -0,0 +1,186 @@
+        // @ts-ignore
+        await page.waitForSelector('.theia-preload', { state: 'detached' });
+        // This selector is for the problems indicator in the status bar, as this is one of the last elements to be rendered.
+        // @ts-ignore
+        await page.waitForSelector('.fa-exclamation-triangle', { state: 'visible' });
```
Shouldn't this be `visible: true`?
```suggestion
        // @ts-ignore
        await page.waitForSelector('.fa-exclamation-triangle', { visible: true });
```

sdirix commented on scripts/performance/README.md:
```diff
@@ -0,0 +1,30 @@
+The script accepts the following optional parameters:
+
+-   `--name`: Specify a name for the current measurement (default: `Measurement`)
+-   `--url`: Point Theia to a url for example for specifying a specifc workspace (default: `http://localhost:3000/#/<pathToMeasurementScript>/workspace>`)
```
```suggestion
-   `--url`: Point Theia to a url for example for specifying a specifc workspace (default: `http://localhost:3000/#/<pathToMeasurementScript>/workspace`)
```

sdirix commented on scripts/performance/README.md:
```diff
@@ -0,0 +1,30 @@
+
+-   `--name`: Specify a name for the current measurement (default: `Measurement`)
+-   `--url`: Point Theia to a url for example for specifying a specifc workspace (default: `http://localhost:3000/#/<pathToMeasurementScript>/workspace>`)
+-   `--folder`: Folder name for the generated tracing files in the `profiles` folder (default: `profiles/profile`)
```
The actual value of the parameter is `profile`
```suggestion
-   `--folder`: Folder name for the generated tracing files in the `profiles` folder (default: `profile`)
```

sdirix reviewed (changes requested):
Two minor comments left

sdirix reviewed (changes requested):
We should also add this to the changelog with a `[scripts]` prefix, e.g. "Added startup performance measurement script"

sgraband commented:
Rebased and added the changelog entry (PR number missing, as we do not have it yet)

sdirix commented on CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 
  - [workspace] added support for multiple selections in 'Add folder to workspace' dialog. [#9684](https://github.com/eclipse-theia/theia/pull/9684)
  - [debug] added support for managing debug sessions for extensions from debug panel (previously only possible using `Hosted Plugin` commands) [#8706](https://github.com/eclipse-theia/theia/pull/8706)
+ - [scripts] added startup performance measurement script [#](https://github.com/eclipse-theia/theia/pull/)
```
This is missing the Contributed by ST message

sdirix reviewed (changes requested):
Please also adapt your commit message to fit to the new PR message.

sdirix commented:
Merged upstream with https://github.com/eclipse-theia/theia/pull/9777
